File: decorator -pattern/java/src/ShadowShapeDecorator.java
package src;

// 具体装饰器2
public class ShadowShapeDecorator extends ShapeDecorator {

  public ShadowShapeDecorator(Shape decoratedShape) {
    super(decoratedShape);
  }

  @Override
  public void draw() {
    // decoratedShape.draw();
    setShadow(decoratedShape);
  }

  private void setShadow(Shape decoratedShape) {
    System.out.println(
      "ShadowShapeDecorator::setShadow() " + decoratedShape.getClass().getName()
    );
  }
}

File: decorator -pattern/python/src/ShadowShapeDecorator.py
# -*- coding: utf-8 -*-
"""
@author: jarry
"""
from src.ShapeDecorator import ShapeDecorator
# 具体装饰器2


class ShadowShapeDecorator(ShapeDecorator):
    def __init__(self, decorated_shape):
        ShapeDecorator.__init__(self, decorated_shape)

    def draw(self):
        # self.decorated_shape.draw();
        self.set_shadow(self.decorated_shape)

    def set_shadow(self, decorated_shape):
        print(
            'ShadowShapeDecorator::set_shadow() ' + decorated_shape.__class__.__name__
        )
